### PR TITLE
File system virtualization proposal

### PR DESCRIPTION
**Pull request by albertofustinoni**
@twinaphex

This is my proposal for handling file access virtualization in a way that both:

- Allows Libretro cores to do things like loading content from archives/streams/whatever by updating just the front end and not every single core
- Allows new platforms with completely different ways to access data (like WinRT) to be supported by updating just the front end

To do so, I am building up from the approach taken in `libretro_common` `filestream`:

- `filestream_transforms` is mostly untouched and keeps its function of redirecting IO access from parts of the core codebase that need to stay compatible with upstream with minimal changes (just ` #include` the header)
- `filestream` handles implementing high order file IO functionality (think `fprintf`) using call-backs to the front end to manipulate file data
- `libretro.h` gains a set of call-backs to be implemented by the front end for data access. The front end is free to interpret paths passed to it as it pleases to generate `RFILES`. These call-backs only deal with basic IO operations to keep their number small and stable: complex data access operations (again, like `fprintf`) are implemented in `filestream_transforms`
- `RFILE` itself becomes an opaque handle to the core (only ever forward declared there), just like `FILE` is: core code and ` filestream_transforms` only ever operates on `RFILE` pointers. This is the exact same approach used by the C language for dealing with `FILE` handles.
- To help with migration to virtualized file access, `filestream` implements fall-back file access functions (in `filestream_fallbacks`, they are actually the same code as in the original `filestream`): if the front end does not set call-backs for file access, that code path is used instead.
Front ends using an old `libretro.h` will keep working with updated cores. Also, you can copy/paste those fall-back functions into RetroArch (changing types as needed) as a starting point for file access support via the front end.
Ideally you would get rid of those fall-back functions once Retroarch is ready.

Some extra things I did:

- Cleaned up the `filestream_transforms` API a bit: `filestream_open` had an extra unused parameters at the end which I removed.
- Had `filestream_transforms` use `int64_t` instead of `ssize_t`: this is both to avoid including headers from `libretro_common` in `libretro.h` and because it makes sense to use 64bit indices for file operations in all cases.

What do you think?

### CODE REVIEW

**Comment by ekeeke** [comment]
Not sure why Genesis Plus GX needs this since it already handles CD file access virtualization by itself in the core (and ROM file access is only handled externally to the core) but /frontend should be moved into /libretro since it's specific to libretro port

**Comment by albertofustinoni** [comment]
@ekeeke This work builds on top of the CD file aces virtualization you included and changes *nothing* of the code upstream.

Look at the changelist: all files modified are in /libretro and subfolders.

I am using your core for the proposal instead of creating a PR for `libretro_common` to:

1. Demonstrate that my proposal requires no changes for cores already using `filestream`
2. Have a chance to discover issues that I could miss if I worked on `libretro_common` code in isolation

**Comment by ekeeke** [comment]
Sorry, I wasn't looking at the right place. I see this was indeed moved into libretro-common, my bad.

**Comment by Alcaro** [comment]
Whining time!

- Everything else is prefixed with retro_, should do that with struct RFILE and that enum too probably. If you want RFILE, maybe put 'typedef retro_vfs_file RFILE;' in the -common header.
- Document what happens if you try to delete a file that doesn't exist. I vote return int; 0 for still there, 1 for removed, -1 for was never there. That way, you can go 'if (vfs->delete(file)) { doesn't exist }'.
- Document the version where each of those were added. It's 1 for everything right now, but we'll need the versions at some point.
- Document what READ_TEXT mode is, and especially why there's no WRITE_TEXT. Or make it a bitflag to OR with the others. If that would break the -common API, maybe stick a translator in filestream_open(), maybe do another PR to remove it prior to landing this. Or maybe just remove right now, apparently we're already changing the API.
- retro_vfs_file_error_t claims to be 'Same as ferror()'. In libc, the ferror flag is cleared by clearerr, rewind and freopen - none of which exist here. All of these return success flags (or something that can be used as success flag) already, maybe it can be removed? If not, at least document what exactly sets and clears the error flag.
- Document whether stuff is deleted if you open an existing file as WRITE. Document whether READ_WRITE creates file or fails if that filename doesn't exist.
- Does get_path return the exact path it was opened with, or is get_path(open(/home/foo/../bar/file.txt)) allowed to return /home/bar/file.txt? And, more importantly, is it useful for anything?
- Document return value from close_t and flush_t.
- Other structs and enums in libretro.h aren't typedefs, only function pointers and some int64_t aliases are. If you want to change that, make a separate PR; for now, let's stay consistent.
- No point having retro_vfs_file prefix on the members of vfs_interface. The retro_ prefixes are useful for anything in the global scope, but that's not the global scope.
- Comments in struct retro_vfs_interface_info refer to a 'requested_interface_version' member that was either renamed or merged into required_.
- Missing docs of which slashes are allowed in paths on Windows. Should probably allow both, and require that front can deal with both.
- Missing docs of what cores may do with paths (extract and print basename, replace basename with file.txt, replace basename with foo/bar.txt, replace basename with foo//bar.txt, replace basename with ./bar.txt, remove the last subdirectory component, replace basename with ../bar.txt, make up its own paths, etc).
- Should document what kinds of paths front may hand out, too. This should probably explicitly permit anything not forbidden by the above, for example 'front may send paths in any format it wants, for example ./mario.sfc, /home/admin/mario.sfc, http://example.com/mario.sfc, #game/mario.sfc'.
- Missing docs of where the main GET_VFS env should be called. It must be called before core sees any other path, which means before retro_load_game and the GET_*_DIRECTORY envs. Might as well say it goes in retro_set_environment, many other envs belong there already.
- On some platforms, rarch is statically linked with the core - and they share one copy of -common. This may give funny results; depending on how the frontend is implemented, it could be 'filestream-using cores get VFS support even if they doesn't ask for it', 'VFS can't be enabled under this configuration', 'infinite recursion and a crash', or something even weirder. Someone better test that. I said on IRC I don't know how to solve that one, but I changed my mind: it can probably be solved with some variant of #ifndef VFS_TARGET_IS_FRONTEND #define filestream_open filestream_open_from_core.
- I've got several objections against the filestream implementation, for example #define MODE_STR_READ "r" #define MODE_STR_READ_UNBUF "rb". But unlike libretro.h, those parts were already weird, and they're way easier to change than libretro. We can worry about that later.

**Comment by albertofustinoni** [comment]
@Alcaro 

Worked on the PR some more to address your complaints.

Significant changes:
- file_error() is out of the VFS API and is handled by filestream in common instead
- Reworked file access flags to be saner. Tell me if they make sense to you now.
- Implementation of file open needs to be updated to conform to new spec -waiting till fe agree on something before doing that.

**Comment by Alcaro** [comment]
That's a lot better.

> Everything else is prefixed with retro_, should do that with struct RFILE and that enum too probably. If you want RFILE, maybe put 'typedef retro_vfs_file RFILE;' in the -common header.

Yes, like that.

> Document what happens if you try to delete a file that doesn't exist. I vote return int; 0 for still there, 1 for removed, -1 for was never there. That way, you can go 'if (vfs->delete(file)) { doesn't exist }'.

Missed this one?

> Document the version where each of those were added. It's 1 for everything right now, but we'll need the versions at some point.

...that's slightly more intrusive than I expected, but it is what I asked for. Good enough. We could clean them up later if they get too annoying, that kind of comments don't constitute an ABI.

> Document what READ_TEXT mode is, and especially why there's no WRITE_TEXT. Or make it a bitflag to OR with the others. If that would break the -common API, maybe stick a translator in filestream_open(), maybe do another PR to remove it prior to landing this. Or maybe just remove right now, apparently we're already changing the API.

I'm still not sure if text mode is needed at all. All cores I'm aware of already account for stray \r; if they didn't, cores on Linux couldn't read files created on Windows. Probably safer to exclude it; worst case, we can put it in v2.

> retro_vfs_file_error_t claims to be 'Same as ferror()'. In libc, the ferror flag is cleared by clearerr, rewind and freopen - none of which exist here. All of these return success flags (or something that can be used as success flag) already, maybe it can be removed? If not, at least document what exactly sets and clears the error flag.

Removal is simple but effective.

> Document whether stuff is deleted if you open an existing file as WRITE. Document whether READ_WRITE creates file or fails if that filename doesn't exist.

Those names seem slightly bulky, but (other than the above complaint about text mode) I am satisfied.

> Does get_path return the exact path it was opened with, or is get_path(open(/home/foo/../bar/file.txt)) allowed to return /home/bar/file.txt? And, more importantly, is it useful for anything?

You forgot the latter half, but okay, good enough.

> Document return value from close_t and flush_t.

Okay, but why does flush return 0 on success, when delete returns true aka 1? And what does failure from those mean?

> Other structs and enums in libretro.h aren't typedefs, only function pointers and some int64_t aliases are. If you want to change that, make a separate PR; for now, let's stay consistent.

You forgot retro_vfs_interface_info.

> No point having retro_vfs_file prefix on the members of vfs_interface. The retro_ prefixes are useful for anything in the global scope, but that's not the global scope.

file_ prefix isn't needed either, that information is already available in the pointer used to access that struct.

> Comments in struct retro_vfs_interface_info refer to a 'requested_interface_version' member that was either renamed or merged into required_.

Yes, like that. Simple typo, easy fix.

> - Missing docs of which slashes are allowed in paths on Windows. Should probably allow both, and require that front can deal with both.
> - Missing docs of what cores may do with paths (extract and print basename, replace basename with file.txt, replace basename with foo/bar.txt, replace basename with foo//bar.txt, replace basename with ./bar.txt, remove the last subdirectory component, replace basename with ../bar.txt, make up its own paths, etc).
> - Should document what kinds of paths front may hand out, too. This should probably explicitly permit anything not forbidden by the above, for example 'front may send paths in any format it wants, for example ./mario.sfc, /home/admin/mario.sfc, http://example.com/mario.sfc, #game/mario.sfc'.

Did you miss these?

> Missing docs of where the main GET_VFS env should be called. It must be called before core sees any other path, which means before retro_load_game and the GET_*_DIRECTORY envs. Might as well say it goes in retro_set_environment, many other envs belong there already.

Should mention the GET_*_DIR envs, maybe? Or can we file that under common sense?

> On some platforms, rarch is statically linked with the core - and they share one copy of -common. This may give funny results; depending on how the frontend is implemented, it could be 'filestream-using cores get VFS support even if they doesn't ask for it', 'VFS can't be enabled under this configuration', 'infinite recursion and a crash', or something even weirder. Someone better test that. I said on IRC I don't know how to solve that one, but I changed my mind: it can probably be solved with some variant of #ifndef VFS_TARGET_IS_FRONTEND #define filestream_open filestream_open_from_core.

Seems unsolved, but it's impossible to judge the severity of this one (let alone solve it) without seeing the frontend code. Let's ignore it for now.

> I've got several objections against the filestream implementation, for example #define MODE_STR_READ "r" #define MODE_STR_READ_UNBUF "rb". But unlike libretro.h, those parts were already weird, and they're way easier to change than libretro.h. We can worry about that later.

It's still weird, but moving weirdness around doesn't make it worse. I'm fine with ignoring this one.


Current issues:
- Missing documentation for what happens if you try to delete a file that doesn't exist
- Is text mode really needed? Text mode converts \r\n to \n on Windows (and opposite for writing), which would be fairly annoying for the fronts to deal with (especially combined with seeking) - and as far as I know, all cores deal with \r already (text mode is a no-op on Linux, so if they want to read Windows files, they have to implement that themselves), so it'd be a lot of effort for no benefit.
- Why do flush() and close() return 0 on success, when delete() returns true aka 1? Let's stick with failure=0 for everything.
- What does flush() or close() failure mean, and how is core expected to react? Should it retry that operation, close handle and try writing the file again, or something weirder?
- struct retro_vfs_interface_info is still a typedef
- file_ prefix is better than retro_vfs_file_, but neither of them is needed on struct members. ...waitno, it is, delete is a keyword in C++, so renaming to delete() would fail. May be better to rename to unlink(), but the current one works too.
- Missing docs of what liberties and requirements core and front have with paths
- Mention GET_*_DIR in the docs for when to call GET_VFS env
- Should mention what front is expected to do if it supports VFS v1 and core demands v2. I vote same as if front doesn't support VFS at all: return false from the env without changing iface.
- Considering the size of this thing, it's probably better to mark this thing experimental for now (RETRO_ENVIRONMENT_EXPERIMENTAL). That way, we could land this feature immediately, then aggressively increment the version if we run into issues.
- Someone better review and/or test stuff outside libretro.h, I only glanced over that. While I found some bad stuff (your usage of that env in libretro.c looks quite suspicious), I probably missed something important.
- Do we have a functional rarch implementation, or a volunteer for creating that?

**Comment by albertofustinoni** [comment]
@Alcaro 

I have made some more changes to address the issues you brought up.

A couple of points I'd like to raise:

- Should we explicitly tell the core the file doesn't exist when deleting? [remove](http://www.cplusplus.com/reference/cstdio/remove/) itself only returns success or failure.
- I don't know if text mode is needed or not, only that it is included in the current `filestream` implementation. If you are confident it won't cause issues I'd be more than glad to get rid of it.
- Changed `delete` to behave like `flush` and `close`. I am using the standard C file IO API as a model for the VFS API, since it is meant to be a drop in replacement to `fopen` etc via `filestream`.
- I think failure when flushing should mean the core should retry. Closing should behave like `fclose` and make the file handle invalid - the core should not try to anything more.

What do you think?

**Comment by Alcaro** [comment]
(are we two the only ones who care about this thing? Twinaphex said he's seen a lot of questions about VFS, where did they go?)

> Should we explicitly tell the core the file doesn't exist when deleting? remove itself only returns success or failure.

remove() itself only returns success or failure, as does unlink() on Unix, and DeleteFile() on Windows.

But all of them set errno, which can be inspected and used to check if file still exists.

But on the other hand, there are few usecases for such a return value, and all plausible usecases can easily be followed by trying to open that file. Probably easier to stick to 0/-1. (And sometimes you don't even know if the file exists, for example if you don't have access to that directory.)

> I don't know if text mode is needed or not, only that it is included in the current filestream implementation. If you are confident it won't cause issues I'd be more than glad to get rid of it.

I'm not confident it's 100% painfree, there may be a few cores that use it.

However, I am confident that all cores using it either don't need it, or are buggy on non-Windows and should be fixed. (Unless we have any Windows-only cores, but platform dependencies are bugs as well.)

> Changed delete to behave like flush and close. I am using the standard C file IO API as a model for the VFS API, since it is meant to be a drop in replacement to fopen etc via filestream.

Not sure if that's the best model, but on the other hand, those RFILE wrappers around the filestreams are indeed modeled after libc, and I think they're supposed to be quite transparent.

No option is perfect. This one is good enough.

Though if you use 0 for success, you should probably use only negatives for failure. if (foo() < 0) handle_error() is a fairly common pattern, let's keep it functional.

> I think failure when flushing should mean the core should retry. Closing should behave like fclose and make the file handle invalid - the core should not try to anything more.

Agreed, close() should always close it. Failure means the implicit flush() failed, nothing else.

Core can choose whether to retry or fail if flush fails. Probably latter, why would it work better this time? (Though I suspect most cores will just assume it always works, leaving old or corrupt data if it fails. Can't fix stupid...)

> File paths passed as parameters when using this api shall be:

Front will get relative or Windows-style paths from the command line, which (using that ruleset) means front needs to convert them. Fronts not supporting VFS will, obviously, not do this, so core still needs to deal with relatives and backslashes (or declare a hard dependency on VFS, but I suspect that wouldn't be very popular).

I fully agree that paths with backslashes are corrupt, but that's the world we live in. We'll just have to deal with it.

The ban on relative paths seems fairly hard to define, too; to allow paths like http://example.com/mario.sfc, we have to allow multiple roots (or some equivalent concept), and then you can define another root named . that just accidentally happens to be equivalent to the current working directory. 

Therefore, let's allow all paths fopen() does. It's not perfect, but if cores still have to handle weird paths on non-VFS, banning them from VFS gains us nothing.

> Cores should treat paths received from the core as they would if there was no VFS in place

Ignoring the typo, let's not say "cores are allowed to do whatever works on the native filesystem", that invites disaster when we try to run cores on weirder platforms, like Xbox or Mac OS Classic. Let's spell out the exact rules.

Unfortunately, I'm under a contradictory set of requirements to find such a ruleset; I don't want to (1) break existing cores, (2) allow absurd paths like http://example.com/derp\mario.sfc, (3) require additional code for the sake of this VFS, (4) cause cores to fail on weird platforms like Xbox, nor (5) require constant changes if we want to port to even weirder platforms, like Mac OS Classic.

The best I can do is relaxing (2), satisfying the other four:

libretro.h contains #define RETRO_DIRSEP '\\' #define RETRO_DIRSEP_ALT '/' (or both '/' if _WIN32 is not defined; we'll add more branches to the ifdefs later if we want a Mac OS Classic port, but that's out of scope for this PR). The name is not RETRO_VFS_DIRECTORY_SEPARATOR because it will be used often, and repeatedly typing long names will get really annoying really fast. There is also #define RETRO_DIRSEP_S "\\", to allow constructing paths at compile time.

To determine the basename, cores must accept both DIRSEP and DIRSEP_ALT. Core is allowed to replace the basename, enter a sibling subdirectory, trim off components from the end, or all of the above, as long as it can deal with front rejecting the new path.

Cores should use DIRSEP if possible (i.e. if not copied verbatim from an external file), as DIRSEP_ALT is '/' on all platforms, including those where it doesn't work in native paths (like Xbox); it must be, or we end up with http:::example.com:mario.sfc on Mac9. Front is allowed to rewrite paths prior to opening, to accomodate cores reading paths from files, as long as get_path returns the original.

Cores are allowed to use #if DIRSEP != DIRSEP_ALT to remove redundancies, if applicable. Cores are not allowed to hardcode DIRSEP, as its definition will change if we port to new platforms; if RETRO_DIRSEP is a too long name, the recommended solution is #define SEP RETRO_DIRSEP. Cores are allowed to hardcode the assumption that DIRSEP_ALT is '/', but for symmetry, this isn't recommended.

Cores are strongly discouraged from making up their own paths (paths not created by replacing components of a front-provided path), due to the large variance between platforms on what the root directory/ies are (some windows boxes don't have a C:\, and consoles use even weirder schemes). If cores do make up paths, they must be absolute, native file system paths; spontaneously loading http://example.com/mario.sfc or ./mario.sfc is strictly forbidden. Frontends are not required to allow native file system access.

> Set by core: should this be higher than the version the front end supports, front end will: 1. set the iface pointer in this struct to NULL

I'd prefer not setting that one at all if that version is unsupported (front will do that if it doesn't acknowledge VFS at all), but okay, both work. Most properly written cores will initialize that field to NULL anyways, changing from NULL to NULL tends to not be a very significant change. (I could see someone wanting to initialize that pointer to some default handler and then using whatever is in that struct after calling the env, but losing that convenience is no big deal. It's not an improvement if using the -common filestreams, anyways.)

Current unresolved issues:
- Return value from delete()
- (I'll assume you decided file_delete is a better name than unlink, as opposed to missing that one)
- Text mode
- Valid path formats
- Handling of unsupported VFS versions
- Reviewing and testing stuff outside libretro.h
- RetroArch implementation

**Comment by albertofustinoni** [comment]
Yeah, at this point it does really feel like we are the only ones who care about this... I tried reaching @Garbear since he started the other VFS PR and XBMC also needs it but no answer.

I addressed some of the points you brought up. About the remaining ones:

- Updated `file_delete` to only return 0 on success or -1 on failure
- I am going to trust your judgement on removing text mode from the API and have done so
- As far as I can tell, the API only returns 0 (success) or -1 (failure) right now
- Documented that handles passed to close become invalid even when the call fails
- Changed handling of mismatched `required_interface_version` as per your suggestion, front end now only returns false to API call.
- I like file_delete better than alternatives, and left the file_ prefix on the other `retro_vfs_interface` members for consistency.

As for paths, I think the best course of action is to just say that cores should not make up their own paths if they want a reasonable expectation of actually getting those files to open (just as things work now). Take one of the paths sent by the front end (either via env callbacks or `retro_load_game`) and modify from there.

There's no reason to outright forbid a core from trying to open `lol#this:"'is??an==insane_path`. If the front end has that mapped to some data it will open, otherwise it won't.
Conceptually the VFS is just a mapping between arbitrary strings and handles to data.

With regards to Retroarch implementation, you will be able to do something like

```
#include <vfs/vfs_implementation.h>

struct retro_vfs_interface iface;
iface.file_open = retro_vfs_file_open_impl;
etc.
```

if you #define `VFS_TARGET_IS_FRONTEND` (can be changed to some other #define already exclusive to Retroarch).

This allows Retroarch to provide an implementation for the VFS API using the same code as the one filestream would use no VFS callbacks are set.

When Retroarch decides what they want to use the VFS for, they'll be able to create a different implementation, but for the time being it's going to make it easier to find and fix bugs in the actual file IO code (which again, is copy/pasted from the original filestream as much as possible to reduce bugs).

As for the actual code review and testing, hopefully someone more knowledgeable/involved in Retroarch can help. I do not have a way to build, let alone test, for all platforms it supports, and of course I cannot use RetriX to test that code (I need the VFS precisely because I cannot use fopen etc.).

**Comment by Alcaro** [comment]
> There's no reason to outright forbid a core from trying to open lol#this:"'is??an==insane_path.

There's no reason to allow it either.

But yeah, we can't really stop them. Let's just not mention http paths. "Cores are allowed to make up paths, but due to large variance between platforms on what the root directory/ies are, this is strongly discouraged, and the core may not rely on any such path working. The frontend is only required to accept the paths it hands out."

> With regards to Retroarch implementation, you will be able to do something like

Yes, but someone has to do it.

And I feel it'd be better to include something more than native files, feels kinda underwhelming to push this much code without actually providing any results. But I'm not sure how necessary that is for the initial implementation.

> I do not have a way to build, let alone test, for all platforms it supports

Not needed. We need to test one platform using dynamic linking, and one platform using static link, but anything beyond that is unlikely to reveal anything interesting. If we allow both models on the desktop OSes (I think we do), an ordinary Win10 desktop would be perfectly sufficient.

Maybe I'll do it. But I'm super lazy, so don't expect it anytime soon.

> of course I cannot use RetriX to test that code

Yes, but you can use RetriX to test the core, and you can test RetriX itself. If it doesn't work for you, I doubt it'd work for anyone else either.

> front end will return false in the RETRO_ENVIRONMENT_GET_VFS_INTERFACE call

Should probably explicitly mention that iface should be left alone. May be pointless, but better verbose than someone misreading the spec.

> Frontend writes interface pointer here.

Prefix with 'on success', maybe.

**Comment by Alcaro** [comment]
> Unfortunately, I'm under a contradictory set of requirements to find such a ruleset; I don't want to (1) break existing cores, (2) allow absurd paths like http://example.com/derp\mario.sfc, (3) require additional code for the sake of this VFS, (4) cause cores to fail on weird platforms like Xbox, nor (5) require constant changes if we want to port to even weirder platforms, like Mac OS Classic.

I just realized something:

Even if not using VFS, path access still goes through the -common filestream wrappers, which will translate pathnames (if necessary) prior to passing them to the OS. That way, we can treat / as the separator everywhere, with only a few caveats:
- If the core renders the paths to the screen, a Windows user expecting backslashes could be confused.
- - Doubt this matters. It doesn't affect functionality, and / is close enough to \\ that everyone except smartasses will know what you mean.
- Non-VFS-aware fronts could still pass paths to the core that includes backslashes.
- - For cores, this is harmless for most uses (opening the path, or in case of the GET_*_DIRECTORY envs, appending stuff); the -common translator will replace / with \\, which doesn't affect existing backslashes. For weirder cases, most cores contain suitable ifdefs already; if that's not enough, we'll add path translators to -common. It's a little extra code, but I doubt any plausible core would need more than ~20 extra lines for that.
- - For fronts, adding this translator everywhere would be a little extra code, but probably not more than ~50 lines even for a front implementing every optional and deprecated part of libretro. It's not perfect, but it's good enough.
- If the core wants -common as an optional dependency, following this proposal may end up passing forward slashes to Windows path handlers. For example, some cores allow being built as standalone, and may want to remove everything libretro-related in that case.
- - Doubt this matters either, anything standalone probably doesn't care about anything except desktops, iPhone and Android - all of which understand forward slashes. Even if they do support Xbox, the solution is replacing #ifdef _WIN32 with #if defined(_WIN32) && !defined(__LIBRETRO__). Or embracing the -common.

tldr: Let's go for forward slashes everywhere. No ifdefs in libretro.h, let filestream_open() deal with those platform differences.

**Comment by albertofustinoni** [comment]
Given what you mentioned in your last comment, are you happy with the current wording of the spec regarding file paths?

```
File paths:
File paths passed as parameters when using this api shall be:
1. Well formed unix-style (so use / instead of \ as directory separator)
2. Absolute: including ".", ".." or "~" as special directory signifiers is not allowed
Cores should treat paths received from the core as they would if there was no VFS in place
```

If not, what would you like to see changed?

**Comment by Alcaro** [comment]
- The ~ symbol carries no significance in any OS API, it's parsed by the shell. Let's not mention it.
- Typo: As far as we know, cores don't receive paths from the core. You mean frontend.
- Either -common gains a function to absoluteify a path (which would be a bunch of platform-specific code I don't think anyone wants to write), or cores must tolerate paths starting with a dot. Latter makes more sense to me, cores must deal with weird prefixes (like http://) anyways.
- Let's make everything as clear as possible. I still don't like the "as they would if there was no VFS in place" part.

Proposed replacement:

>File paths:
>File paths passed as parameters when using this api shall be well formed unix-style, using / as directory separator, even if the native separator is something else.
>Other than what the directory separator is, cores may not make any assumptions on how a path looks; paths may be prefixed with whatever the frontend wants. For example, the following paths are valid: C:/path/game.bin, http://example.com/game.bin, #game/game.bin, game.bin
>Cores may replace the basename or remove path components from the end, and/or add new components (including . and/or ..). The frontend is encouraged to make such paths work as well as it can, but is allowed to give up if the core alters paths too much.
>Frontends are encouraged, but not required, to support native file system paths (modulo replacing the directory separator, if applicable). Cores are allowed to try this, but must remain functional if the front rejects such a request.
>Since paths using / don't natively work on all operating systems, cores are encouraged to use the libretro-common filestream functions, which will seamlessly integrate with VFS, and deal with platform differences if VFS is not available.

I'd prefer if I could make it shorter, but I can't find anything useless. Better too many details than too few.

**Comment by albertofustinoni** [comment]
I have edited the spec to reflect your suggestions about file paths.

Two points I want to bring up:

> Frontends are encouraged, but not required, to support native file system paths... Cores are allowed to try this, but must remain functional if the front rejects such a request.
 
So cores need to support VFS style paths anyways - can we just remove this part and say cores who opt in to the VFS need to use VFS style paths? Easier to implement for front ends and cores lose nothing.

> Cores may replace the basename or remove path components from the end, and/or add new components (including . and/or ..)

This is going to make front end implementation a lot more complex: what we are promising here is that cores can send

- /path/to/file.bin
- /path/to/./././file.bin
- /path/lol/../to/file.bin

And the front will need to provide a handle to the same data.

This means that any front end will have to implement some non trivial parsing logic to translate paths requested by cores into a canonical representation (an absolute path), as opposed to, say, use a simple hash table, in order to service requests.

Keep in mind that front ends will not always be able to rely on the OS to handle paths for them, especially if they want to use the VFS for interesting things (mapping http URIs to virtual file paths for example).

Is the increase in complexity worth it? What are the use cases for allowing this instead of having cores generate absolute paths to begin with?

**Comment by Alcaro** [comment]
> So cores need to support VFS style paths anyways - can we just remove this part and say cores who opt in to the VFS need to use VFS style paths? Easier to implement for front ends and cores lose nothing.

"modulo replacing the directory separator, if applicable" means forward slash remains the only allowed one, C:\path\game.bin is still illegal.

Almost-native paths like C:/path/game.bin already work with the default -common implementation, and for fronts who don't want to do that, "Frontends are encouraged, but not required" says ignoring that is fine. Doesn't get any easier than that.

And cores do lose something: as I've said a few times, a few cores (like Dosbox) want to navigate the entire file system. Even if we don't have a directory listing function yet, we most likely will at some point. Even before that, cores could call FindFirstFile or whatever and pass that to the VFS.

> what we are promising here is that cores can send [...] And the front will need to provide a handle to the same data.

Some programs, for example GCC, rely on ../ components already. Someone needs to parse them, whether front or core. (Though they're rare, some don't need to be libretro cores, and some of them are probably easy to change.)

The obvious solution would be sticking that in the VFS-to-native function in -common, but that'd yield http:\\\\example.com\game.bin, which helps nobody. It would also act weirdly if front sends C:/path/../path/path/path/path/game.bin, core isn't allowed to mess with any path component without messing with all subsequent ones. (Fronts relying on ../ components remaining are probably buggy, but "works on any sane front even if it's not allowed per the spec" is still a spec violation and a core bug.)

Instead, let's use a slightly less obvious solution: Create a resolve-path function in -common (resolve("C:/path/game.bin", "../path2/game.bin") -> "C:/path2/game.bin") and tell cores to use that. Cores are not allowed to append ./, ../ or duplicate slashes (C:/path//game.bin) and pass that to the front.

...which is pretty much exactly your proposal.

However, I do propose another adjustment: A path must contain at least one slash, "game.bin" is illegal. This allows cores to find basename of all paths via strrchr(path,'/')+1, rather than dealing with strrchr returning NUL. It's no extra complexity for the fronts, either; they already need the -common native-to-vfs function, teaching it to prepend ./ as appropriate is no extra effort for them.

**Comment by albertofustinoni** [comment]
I see your points.

I have updated the wording again to reflect what we discussed.
Do you think it's ok now?
